File: lib/java/src/main/java/io/github/brenbar/uapi/internal/ApplyErrorToParsedTypes.java
package io.github.brenbar.uapi.internal;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

import io.github.brenbar.uapi.UApiSchemaParseError;

public class ApplyErrorToParsedTypes {
    static void applyErrorToParsedTypes(int errorIndex, _UError error, Map<String, _UType> parsedTypes,
            Map<String, Integer> schemaKeysToIndex) {
        var parseFailures = new ArrayList<_SchemaParseFailure>();
        for (var parsedType : parsedTypes.entrySet()) {
            _UFn f;
            try {
                f = (_UFn) parsedType.getValue();
            } catch (ClassCastException e) {
                continue;
            }

            String fnName = f.name;

            var regex = Pattern.compile(f.errorsRegex);

            _UUnion fnResult = f.result;
            Map<String, _UStruct> fnResultCases = fnResult.cases;
            _UUnion errorErrors = error.errors;
            Map<String, _UStruct> errorCases = errorErrors.cases;

            for (var errorCase : errorCases.entrySet()) {
                var newKey = errorCase.getKey();

                var matcher = regex.matcher(newKey);
                if (!matcher.find()) {
                    continue;
                }

                if (fnResultCases.containsKey(newKey)) {
                    final var otherPathIndex = schemaKeysToIndex.get(fnName);
                    final var errorCaseIndex = error.errors.caseIndices.get(newKey);
                    final var fnErrorCaseIndex = f.result.caseIndices.get(newKey);
                    parseFailures.add(new _SchemaParseFailure(List.of(errorIndex, "errors", errorCaseIndex, newKey),
                            "PathCollision", Map.of("other", List.of(otherPathIndex, "->", fnErrorCaseIndex, newKey)),
                            null));
                }
                fnResultCases.put(newKey, errorCase.getValue());
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new UApiSchemaParseError(parseFailures);
        }
    }
}

File: lib/ts/uapi/internal/schema/ApplyErrorToParsedTypes.ts
import { UError } from 'uapi/internal/types/UError';
import { UStruct } from 'uapi/internal/types/UStruct';
import { UType } from 'uapi/internal/types/UType';
import { UUnion } from 'uapi/internal/types/UUnion';
import { SchemaParseFailure } from 'uapi/internal/schema/SchemaParseFailure';
import { UApiSchemaParseError } from 'uapi/UApiSchemaParseError';
import { UFn } from 'uapi/internal/types/UFn';

export function applyErrorToParsedTypes(
    errorIndex: number,
    error: UError,
    parsedTypes: { [key: string]: UType },
    schemaKeysToIndex: { [key: string]: number },
): void {
    const parseFailures: SchemaParseFailure[] = [];

    for (const parsedTypeName in parsedTypes) {
        const parsedType = parsedTypes[parsedTypeName];

        if (!(parsedType instanceof UFn)) {
            continue;
        }

        const f = parsedType;
        const fnName = f.name;
        const regex = new RegExp(f.errorsRegex);
        const fnResult = f.result;
        const fnResultCases = fnResult.cases;
        const errorErrors = error.errors;
        const errorCases = errorErrors.cases;

        for (const errorCaseName in errorCases) {
            const errorCase = errorCases[errorCaseName];
            const newKey = errorCaseName;
            const matcher = regex.exec(newKey);

            if (!matcher) {
                continue;
            }

            if (newKey in fnResultCases) {
                const otherPathIndex = schemaKeysToIndex[fnName];
                const errorCaseIndex = error.errors.caseIndices[newKey];
                const fnErrorCaseIndex = f.result.caseIndices[newKey];
                parseFailures.push(
                    new SchemaParseFailure(
                        [errorIndex, 'errors', errorCaseIndex, newKey],
                        'PathCollision',
                        { other: [otherPathIndex, '->', fnErrorCaseIndex, newKey] },
                        null,
                    ),
                );
            }

            fnResultCases[newKey] = errorCase;
        }
    }

    if (parseFailures.length > 0) {
        throw new UApiSchemaParseError(parseFailures);
    }
}
